diff -ruN a/vm_insnhelper.c b/vm_insnhelper.c
--- a/vm_insnhelper.c	2024-04-23 10:23:26.000000000 +0000
+++ b/vm_insnhelper.c	2025-12-22 07:18:11.357282031 +0000
@@ -4476,6 +4476,17 @@ vm_invoke_ifunc_block(rb_execution_context_t *ec, rb_control_frame_t *reg_cfp,
     CALLER_REMOVE_EMPTY_KW_SPLAT(ec->cfp, calling, ci);
     argc = calling->argc;
-    val = vm_yield_with_cfunc(ec, captured, captured->self, argc, STACK_ADDR_FROM_TOP(argc), calling->kw_splat, calling->block_handler, NULL);
-    POPN(argc); /* TODO: should put before C/yield? */
+    const VALUE *argv = STACK_ADDR_FROM_TOP(argc);
+    /*
+     * In sorbet_ruby we move this POPN above vm_yield_call_with_cfunc, so that
+     * if compiled code for a block uses EC_JUMP_TAG, the state will be in the
+     * expected state.
+     *
+     * We pass "args_on_stack = TRUE" to vm_yield_with_cfunc to push the
+     * initial stack pointer of the callee frame past the arguments, since it
+     * still needs to be able to read them.
+     */
+    POPN(argc);
+    bool args_on_stack = TRUE;
+    val = vm_yield_with_cfunc(ec, captured, captured->self, argc, argv, calling->kw_splat, calling->block_handler, is_lambda, NULL, args_on_stack);
     return val;
 }
